### ISSUE TITLE
Double free or corruption in sr_unsubscribe

### ISSUE THREAD

**Issue by awesomenode** [sysrepo/sysrepo#2338]
Hi,

I encountered an error in `sr_unsubscribe()`:

```
[ERR]: Locking a mutex failed (sr_subs_del_all: Invalid argument).
[ERR]: unlink() failed (No such file or directory).
double free or corruption (!prev)

Thread 1 "test" received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
51	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007ffff75a1921 in __GI_abort () at abort.c:79
#2  0x00007ffff75ea967 in __libc_message (action=action@entry=do_abort, fmt=fmt@entry=0x7ffff7717b0d "%s\n") at ../sysdeps/posix/libc_fatal.c:181
#3  0x00007ffff75f19da in malloc_printerr (str=str@entry=0x7ffff7719838 "double free or corruption (!prev)") at malloc.c:5342
#4  0x00007ffff75f8f7c in _int_free (have_lock=0, p=0x555555774940, av=0x7ffff794cc40 <main_arena>) at malloc.c:4311
#5  __GI___libc_free (mem=0x555555774950) at malloc.c:3134
#6  0x00007ffff7b771ee in _sr_unsubscribe (subscription=0x555555774950) at /opt/dev/sysrepo/src/sysrepo.c:2831
#7  0x00007ffff7b7e31e in sr_unsubscribe (subscription=<optimized out>) at /opt/dev/sysrepo/src/sysrepo.c:2844
#8  0x00005555555557d6 in unsubscribe () at test.c:236
#9  0x0000555555555bed in test_sub () at test.c:341
#10 0x0000555555555d07 in main (argc=2, argv=0x7fffffffdb98) at test.c:377
```

I attached a test to reproduce it.
[test.zip](https://github.com/sysrepo/sysrepo/files/6030042/test.zip)

This test has three processes:
- test init: creates instances and subscribes for interface list
- test set: continuously changes data in running and operational datastore
- test sub: continuously subscribes for data in running and operational datastore, then unsubscribes

```
$ ./test init & sleep 1 ; ./test set & sleep 1 ; ./test sub
[1] 12515
Test init
[2] 12518
Starting set test
Starting sub test
[ERR]: Waiting on a conditional variable failed (sr_shmext_change_subscription_del: Connection timed out).
unsubscribe timed out, retrying
[ERR]: Locking a mutex failed (sr_subs_del_all: Invalid argument).
[ERR]: unlink() failed (No such file or directory).
double free or corruption (!prev)
Aborted (core dumped)
```

First `sr_unsubscribe()` times out in test sub, then it tries again, which results in double free error.

Installed yang modules:
```
ietf-interfaces            | 2018-02-20
iana-if-type               | 2018-07-03
```

sysrepo version: 1.4.117
libyang version: 1.0.221

Thanks!

**Comment by michalvasko** [comment]
Firstly, you have missed the note in `sr_unsubscribe()`, which is causing the memory issues. Also, while trying your example, I found some obsolete code that I removed and which could cause invalid assert failures and skipped events.

As for the actual issue that is left, it is a dead lock of event being generated with the originator waiting for its handling and unsubscribing from it at the same time. Specifically, there is a synchronization problem of handling events but stopping when the subscription was unsubscribed (to avoid using invalid memory) and wanting to unsubscribe but not while there are some events to be processed (to avoid this dead lock).

Also, this problem should be present only when using a custom event loop when *sysrepo* cannot [enforce the operation order](https://github.com/sysrepo/sysrepo/blob/devel/src/sysrepo.c#L2798) mentioned above and when I was fixing it there, I realized the problem is still present for a custom loop. I have some ideas how to solve this but did not have time to implement any of them, I will write here once it is done.
